Drivers/Src/Display.c

```diff
@@ -1,325 +1,141 @@
-
-/**
- * Sends out a string of the form "obj.attr=val" or "attr=val"
- * Use to update integer fields of display objects
- */
-static ErrorStatus updateIntValue(enum CommandString_t obj_index, enum CommandString_t attr_index, int32_t val)
-{
-    sendStartOfAssignment(obj_index, attr_index);
-
-    char number[12]; // To store converted int
-    sprintf(number, "%ld", val);
-
-    BSP_UART_Write(DISP_OUT, number, strlen(number));
-    BSP_UART_Write(DISP_OUT, (char *)TERMINATOR, strlen(TERMINATOR));
+#include "Display.h"
+#include "bsp.h"   // for writing to UART
+#include "Tasks.h" // for os and fault error codes
 
-    // Get a response from the display
-    char buf[8];
-    BSP_UART_Read(DISP_OUT, buf);
-    return (IsNextionFailure(buf[0])) ? ERROR : SUCCESS;
-}
+#define DISP_OUT UART_3
+static const char *TERMINATOR = "\xff\xff\xff";
 
-/*
- * Sets a component's visiblity
- */
-ErrorStatus setComponentVisibility(enum CommandString_t comp, bool vis)
+DisplayError_e Display_Init()
 {
-    char out[24];
-    sprintf(out, "%s %s,%d%s", CommandStrings[VIS], CommandStrings[comp], vis ? 1 : 0, TERMINATOR);
-
-    BSP_UART_Write(DISP_OUT, out, strlen(out));
-
-    char buf[8];
-    BSP_UART_Read(DISP_OUT, buf);
-    return (IsNextionFailure(buf[0]) ? ERROR : SUCCESS);
+	BSP_UART_Init(DISP_OUT);
+	return Display_Reset();
 }
 
-/*
- * Initialize the Nextion display
- */
-void Display_Init()
+DisplayError_e Display_Send(DisplayCmd_t cmd)
 {
-    char ret[8];
-    for (int i = 0; i < 8; i++)
-        ret[i] = 0;
-    volatile char *x = ret;
-    BSP_UART_Init(DISP_OUT);
-
-    // The display sends 0x88 when ready, but that might be
-    // before we initialize our UART
-    BSP_UART_Read(DISP_OUT, (char *)x);
-    if (ret[0] == 0x88)
-    {
-        volatile int a = 0;
-        while (1)
-            a++;
-    }
-
-    updateIntValue(PAGE, BCO, NEXTION_DARK_GREY);
-}
-
-/**
- * Set the displayed velocity to vel
- */
-ErrorStatus Display_SetVelocity(float vel)
+	char msgArgs[32];
+	if (cmd.compOrCmd != NULL && cmd.op != NULL && cmd.attr != NULL && cmd.numArgs == 1)
+	{ // Assignment commands have only 1 arg, an operator, and an attribute
+		if (cmd.argTypes[0] == INT_ARG)
+		{
+			sprintf(msgArgs, "%d", (int)cmd.args[0].num);
+		}
+		else
+		{
+			if (cmd.args[0].str == NULL)
+				return DISPLAY_ERR_PARSE;
+			sprintf(msgArgs, "%s", cmd.args[0].str);
+		}
+
+		BSP_UART_Write(DISP_OUT, cmd.compOrCmd, strlen(cmd.compOrCmd));
+		BSP_UART_Write(DISP_OUT, ".", 1);
+		BSP_UART_Write(DISP_OUT, cmd.attr, strlen(cmd.attr));
+		BSP_UART_Write(DISP_OUT, cmd.op, strlen(cmd.op));
+	}
+	else if (cmd.op == NULL && cmd.attr == NULL)
+	{					  // Operational commands have no attribute and no operator, just a command and >= 0 arguments
+		msgArgs[0] = ' '; // No args
+		msgArgs[1] = '\0';
+		if (cmd.numArgs > MAX_ARGS)
+			return DISPLAY_ERR_OTHER;
+		if (cmd.numArgs >= 1)
+		{ // If there are arguments
+			for (int i = 0; i < cmd.numArgs; i++)
+			{
+				char arg[16];
```

Comment:
why is this 16 bytes? Magic numbers = bad, use macros or annotate the line with why its 16.